balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+			continue
+		}
+		newEndpoints = append(newEndpoints, resolver.Endpoint{
+			Addresses:  addrs,
+			Attributes: ep.Attributes,
+		})
+	}
+	return newEndpoints
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) reconcileSubConns(newEndpoints []resolver.Endpoint) {
+	// Remove old subConns that were not in new address list.
+	oldAddrs := resolver.NewAddressMap()
+	for _, k := range b.subConns.Keys() {
+		oldAddrs.Set(k, true)
+	}
+
+	// Flatten the new endpoint addresses.
+	newAddrs := resolver.NewAddressMap()
+	for _, endpoint := range newEndpoints {
+		for _, addr := range endpoint.Addresses {
+			newAddrs.Set(addr, true)
+		}
+	}
+
+	// Shut them down and remove them.
+	for _, oldAddr := range oldAddrs.Keys() {
+		if _, ok := newAddrs.Get(oldAddr); ok {
+			continue
+		}
+		val, _ := b.subConns.Get(oldAddr)
+		val.(*scData).subConn.Shutdown()
+		b.subConns.Delete(oldAddr)
+	}
+}
+
+// shutdownRemaining shuts down remaining subConns. Called when a subConn
+// becomes ready, which means that all other subConn must be shutdown.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) shutdownRemaining(selected *scData) {
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.subConn != selected.subConn {
+			sd.subConn.Shutdown()
+		}
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+	b.subConns.Set(selected.addr, selected)
+}
+
+// requestConnection starts connecting on the subchannel corresponding to the
+// current address. If no subchannel exists, one is created. If the current
+// subchannel is in TransientFailure, a connection to the next address is
+// attempted.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) requestConnection() {
+	if !b.addressList.isValid() || b.state == connectivity.Shutdown {
+		return
+	}
+	curAddr := b.addressList.currentAddress()
+	sd, ok := b.subConns.Get(curAddr)
+	if !ok {
+		var err error
+		// We want to assign the new scData to sd from the outer scope, hence
+		// we can't use := below.
+		sd, err = b.newSCData(curAddr)
+		if err != nil {
+			// This should never happen, unless the clientConn is being shut
+			// down.
+			b.logger.Warningf("Failed to create a subConn for address %v: %v", curAddr.String(), err)
+			// The LB policy remains in TRANSIENT_FAILURE until a new resolver
+			// update is received.
+			b.state = connectivity.TransientFailure
+			b.addressList.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("failed to create a new subConn: %v", err)},
+			})
+			return
+		}
+		b.subConns.Set(curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressList.increment() {
+			b.endFirstPass(scd.lastErr)
+			return
+		}
+		b.requestConnection()
```

Comment:
`requestConnection` tries to find the next subconn to connect. It can recurse upto `# of subconns` level deep if all the subconns are in TF. At this point it will end the first pass and return.

The same logic can be written as a for loop that breaks when `b.addressList.increment() == false`, but it adds one layer of indentation. Do you want me to remove the recursive call?

Reply:
Maybe see how it looks the other way?  Otherwise at least add a comment `// Try the next address until we find a subchannel that isn't in transient failure.` (or whatever).